### ISSUE TITLE
Legend change position with responsive rule

### ISSUE THREAD

**Issue by RaulPampliegaMayoral** [amcharts/amcharts4#2347]
Hi!

I want to change the legend position of a Pie chart with small width. I know this can be achieve with responsive rules and this is what i've defined:

````
chart.responsive.enabled = true;
chart.responsive.useDefault = false;

chart.responsive.rules.push({
  relevant: function(target) {
    if (target.pixelWidth <= 600) {
      return true;
    }
    return false;
  },
  state: function(target, stateId) {
    if (target instanceof am4charts.PieSeries) {
                    target.legend.position = "bottom";
                    target.legend.labels.template.maxWidth = 100;
                    target.legend.labels.template.truncate = true;
    }
return null;
});
````

When width is lower than 600px legend is positioned at bottom, but when is higher is not being repositioned (f.e ipad)

Does someone know how can this be achieve?

Thanks

**Comment by martynasma** [comment]
This is not how responsive rules work. The `state` method needs to return actual `SpriteState` object, not set properties on an object. This function is called only once (to create a state object), hence it working only once for you.

You also want to target `Legend`, not `PieSeries`.

```
chart.responsive.rules.push({
  relevant: function(target) {
    if (target.pixelWidth <= 600) {
      return true;
    }
    return false;
  },
  state: function(target, stateId) {
    if (target instanceof am4charts.Legend) {
      var state = target.states.create(stateId);
      state.properties.position = "bottom";
      
      var labelState = target.labels.template.states.create(stateId);
      labelState.properties.maxWidth = 100;
      labelState.properties.truncate = true;
      return state;
    }
    return null;
  }
});
```

More about creating responsive rules:
https://www.amcharts.com/docs/v4/concepts/responsive/#Adding_custom_rules

**Comment by RaulPampliegaMayoral** [comment]
@martynasma 
Thanks! I will try it!

One more question, It is possible to change the position of the legend wit a responsive rule?

Many thanks in advance

**Comment by RaulPampliegaMayoral** [comment]
@martynasma 
Many thanks! Is working like a charm!
i didnt know i could use that property over Legend state. Can i use on SpriteState object all properties defined on object's template (f.e when i want truncate labels for legend i'm doing ````chart.legend.labels.template.truncate = true;````)

**Comment by martynasma** [comment]
When you apply a state on an object (say `Legend`) the same state is applied on its children as well (labels).

So if you look at my code above, I'm creating a state for Legend itself and returning it.

I also create state for legend's labels with the same id. When legend gets that state set (by responsive routine), so get labels, and get truncated.

**Comment by RaulPampliegaMayoral** [comment]
@martynasma 
I dont know why but the maxWidth and truncate properties are not being applied. Any ideas?

![imagen](https://user-images.githubusercontent.com/13700946/79223170-1fdd3000-7e59-11ea-9bc0-0c36de0a22cb.png)

**Comment by martynasma** [comment]
I'd suggest using `disabled = true` instead of setting value label text to an empty string.

**Comment by RaulPampliegaMayoral** [comment]
@martynasma
Again with ````disabled = true```` is working ok.

What i cannot find out a solution is when i rotate ipad. It apply correctly the rule to put legend on bottom but not the maxWidth rule. And when i rotate again the ipad legend is back to right position but without applying maxWidth and truncate.
Do you have any ideas why can this be?

**Comment by martynasma** [comment]
Can you post your current whole chart setup?

**Comment by RaulPampliegaMayoral** [comment]
Hi @martynasma 

I put the configuration as we have it in gwt. It is very similar to typescript and I think it is well understood

````
chart =  Js.uncheckedCast(Am4Core.create(this.div, Am4Charts.PieChart));
chart.data        = params.getValues();
chart.innerRadius = new Percent(60);
chart.radius      = new Percent(100);
chart.legend      = new Legend();
chart.legend.position = "right";
chart.legend.fontSize = 11;
chart.legend.useDefaultMarker = true;
chart.legend.labels.template.truncate = true;
chart.legend.labels.template.maxWidth = 150;
chart.legend.labels.template.tooltipText    = "{category}";
chart.legend.valueLabels.template.align     = "left";
chart.legend.valueLabels.template.textAlign = "end";
chart.legend.markers.template.width   = 20;
chart.legend.markers.template.height  = 20;

Sprite marker = chart.legend.markers.template.children.getIndex(0);
marker.cornerRadius(12, 12, 12, 12);
marker.strokeWidth = 1;
marker.strokeOpacity = 1;

chart.responsive.enabled    = true;
chart.responsive.useDefault = false;

ResponsiveRule ruleMax405 = new ResponsiveRule();
ruleMax405.relevant = target -> target.pixelWidth <= 341;
ruleMax405.state = (target, stateId) -> {
   if( target instanceof Legend)
   {
       Legend legend = Js.uncheckedCast(target);
       SpriteState<SpriteProperties> state = Js.uncheckedCast(legend.states.create(stateId));
       state.properties.disabled = true;
       return state;
    }
            
    return null;
};

ResponsiveRule rule405And550 = new ResponsiveRule();
rule405And550.relevant = target ->  target.pixelWidth > 342 && target.pixelWidth <= 465;
rule405And550.state = (target, stateId) -> {
    if( target instanceof Legend)
    {
        Legend legend = Js.uncheckedCast(target);
        SpriteState<SpriteProperties> state = Js.uncheckedCast(legend.states.create(stateId));
        state.properties.position = "bottom";
                
        SpriteState<SpriteProperties> labelState = legend.labels.template.states.create(stateId);
        labelState.properties.maxWidth = 100;
        labelState.properties.truncate = true;
                
        SpriteState<SpriteProperties> valueLabelState = legend.valueLabels.template.states.create(stateId);
        valueLabelState.properties.disabled = true;

        return state;
    }
            
    return null;
};
        
chart.responsive.rules.push(ruleMax405);
chart.responsive.rules.push(rule405And550);

PieSeries pieSeries                = chart.series.push(new PieSeries());
pieSeries.dataFields.value         = params.getField();
pieSeries.dataFields.category      = params.getCategory();
pieSeries.labels.template.disabled = true;
pieSeries.properties.scale = 1;
            
SpriteState<SliceProperties> hoverState  = pieSeries.slices.template.states.getKey("hover");
hoverState.properties.scale              = 1;

SpriteState<SliceProperties> activeState = pieSeries.slices.template.states.getKey("active");
activeState.properties.shiftRadius       = 0;
````

Also i have just upgrade to the last version (4.9.16) and the legend show incorrect characters when truncates

![imagen](https://user-images.githubusercontent.com/13700946/79646904-5fd14980-81b0-11ea-8507-0d85ea7bcc43.png)

I can use the earlier version we have been using till now (4.5.14). I can file an issue if it is suitable

**Comment by martynasma** [comment]
Thanks. Can you post the data as well?

And yes, please post weird character thing as a separate issue.

**Comment by RaulPampliegaMayoral** [comment]
I've just created a codepen example https://codepen.io/RaulPampliegaMayoral/pen/QWjEBzG

Behavior is not exactly like our development. Also, the weird characters have disappeared. I'm going to look a little more before I continue to bother you.

Thank you so much for everything

**Comment by RaulPampliegaMayoral** [comment]
@martynasma 

Weird character was due to encoding. Once i set up "utf-8" is working ok.

I'm still with problem with the responsive features. If you can see the codepen I posted with 350px maxWidth should be 50 but this is not being applied. May i have something wrong with the configuration?

**Comment by RaulPampliegaMayoral** [comment]
@martynasma 
I have been able to adjust a little bit the responsive feature. Now i wrap legen label and have better results.
But when responsive rule must be applied again is not working well:

![amchart_responsive_error](https://user-images.githubusercontent.com/13700946/81038416-3994f480-8ea6-11ea-90be-3b195b39c81f.gif)

**Comment by martynasma** [comment]
Can you post updated CodePen?
